nucliadb/nucliadb/search/search/find_merge.py

```diff
@@ -0,0 +1,336 @@
+    else:
+        splitted_text = extracted_text.text[start:end]
+
+    if highlight:
+        splitted_text = highlight_paragraph(
+            splitted_text, words=matches, ematches=ematches  # type: ignore
+        )
+
+    return splitted_text
+
+
+async def set_text_value(
+    kbid: str,
+    result_paragraph: TempFindParagraph,
+    highlight: bool = False,
+    ematches: List[str] = [],
+):
+    # TODO: Improve
+    assert result_paragraph.paragraph
+    assert result_paragraph.paragraph.position
+    result_paragraph.paragraph.text = await get_text_find_paragraph(
+        rid=result_paragraph.rid,
+        kbid=kbid,
+        field=result_paragraph.field,
+        start=result_paragraph.paragraph.position.start,
+        end=result_paragraph.paragraph.position.end,
+        split=None,  # TODO
+        highlight=highlight,
+        ematches=ematches,
+        matches=[],  # TODO
+    )
+
+
+async def set_resource_metadata_value(
+    kbid: str,
+    resource: str,
+    show: List[ResourceProperties],
+    field_type_filter: List[FieldTypeName],
+    extracted: List[ExtractedDataTypeName],
+    find_resources: Dict[str, FindResource],
+):
+    serialized_resource = await serialize(
+        kbid,
+        resource,
+        show,
+        field_type_filter=field_type_filter,
+        extracted=extracted,
+        service_name=SERVICE_NAME,
+    )
+    find_resources[resource].parse_obj(serialized_resource)
+
+
+async def fetch_find_metadata(
+    find_resources: Dict[str, FindResource],
+    result_paragraphs: List[TempFindParagraph],
+    kbid: str,
+    show: List[ResourceProperties],
+    field_type_filter: List[FieldTypeName],
+    extracted: List[ExtractedDataTypeName],
+    highlight: bool = False,
+    ematches: List[str] = [],
+):
+    resources = []
+    operations = []
+    for result_paragraph in result_paragraphs:
+        if result_paragraph.paragraph is not None:
+            find_resources.setdefault(
+                result_paragraph.rid, FindResource(id=result_paragraph.rid, fields={})
+            ).fields.setdefault(
+                result_paragraph.field, FindField(paragraphs={})
+            ).paragraphs[
+                result_paragraph.paragraph.id
+            ] = result_paragraph.paragraph
+
+            operations.append(
+                set_text_value(
+                    kbid=kbid,
+                    result_paragraph=result_paragraph,
+                    highlight=highlight,
+                    ematches=ematches,
+                )
+            )
+            resources.append(result_paragraph.rid)
+
+    for resource in set(resources):
+        operations.append(
+            set_resource_metadata_value(
+                kbid=kbid,
+                resource=resource,
+                show=show,
+                field_type_filter=field_type_filter,
+                extracted=extracted,
+                find_resources=find_resources,
+            )
+        )
+    await asyncio.wait(*operations)
```

Comment:
What do you think about limiting the concurrency of these operations with an [asyncio sempahore](
https://docs.python.org/3/library/asyncio-sync.html#asyncio.Semaphore)?